src/utils/futures.rs

```diff
@@ -280,6 +280,39 @@ pub async fn delay(duration: Duration) {
         .ok();
 }
 
+/// Time a future execution and expose using metrics.
+///
+/// The future is expected to resolve to a Result and the status is logged as well.
+pub async fn timed_task<F, T, E>(task_name: &str, future: F) -> F::Output
+where
+    F: Future<Output = Result<T, E>>,
+{
+    let start_time = Instant::now();
+    let ret = future.await;
+    metric!(
+        timer("futures.done") = start_time.elapsed(),
+        "task_name" => task_name,
+        "status" => if ret.is_ok() { "ok" } else {"err"},
+    );
+    ret
+}
+
+pub trait TimedTaskExt<T, E>: Future<Output = Result<T, E>> + Sized {
+    fn task_metrics<'s>(
+        self,
+        task_name: &'s str,
+    ) -> std::pin::Pin<Box<dyn Future<Output = Self::Output> + 's>>
```

Comment:
On first glance, this is rather strange to look at. A few thoughts and ideas:

 - Can't you just return a plain `Box` here? When `.await`'ing, it will be pinned for you, usually. I might be missing something, though.
 - Usually, we would return a named type here. This becomes more complicated when calling an `async` function, of course. But if you prefer this pattern, I would model it more after tokio's [`Stream::timeout`](https://docs.rs/tokio/0.2.21/tokio/stream/trait.StreamExt.html#method.timeout).
- Since you have an async function already, why not just expose it directly? Note that neither `async-std` nor `tokio` do extension traits anymore and prefer plain functions, which support a more imperative coding style. It's not that unergonomic:

```rust
let search = cached_sentry_search(query);
let result = measure("downloads.sentry.index", search).await?;
```

- _nit_: You'll notice that I called the function in imperative language.